File: sample-conference-videochat/sample-conference-videochat/Components/Calls/ConferenceUserCell/ConferenceUserCell.m
//
//  OpponentCollectionViewCell.m
//  sample-conference-videochat
//
//  Created by Injoit on 11.12.14.
//  Copyright (c) 2014 QuickBlox Team. All rights reserved.
//

#import "ConferenceUserCell.h"
#import "UILabel+Chat.h"
#import "UIView+Chat.h"
#import "NSString+Chat.h"

@interface ConferenceUserCell()
@property (weak, nonatomic) IBOutlet UILabel *userNameTopLabel;
@property (weak, nonatomic) IBOutlet UIView *userView;
@property (weak, nonatomic) IBOutlet UILabel *userAvatarLabel;
@property (weak, nonatomic) IBOutlet UIImageView *userAvatarImageView;
@property (weak, nonatomic) IBOutlet UILabel *nameLabel;
@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (weak, nonatomic) IBOutlet UIImageView *unmuteImageView;
@property (weak, nonatomic) IBOutlet UIView *videoContainerView;
@property (weak, nonatomic) IBOutlet UIImageView *unmuteOnVideoImageView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *unmuteImageViewWidthConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *nameLabelCenterXConstraint;

@end

@implementation ConferenceUserCell

- (void)awakeFromNib {
    [super awakeFromNib];
    
    self.isResizeAspect = YES;
    self.videoEnabled = YES;
    self.isMuted = NO;
    self.nameColor = UIColor.clearColor;
    self.backgroundColor = UIColor.clearColor;
    self.videoContainerView.backgroundColor = UIColor.clearColor;
    self.containerView.backgroundColor = [UIColor colorWithRed:0.20f green:0.20f blue:0.20f alpha:1.0f];
    [self.userAvatarLabel setRoundedLabelWithCornerRadius:30.0f];
    [self.userAvatarImageView setRoundViewWithCornerRadius:30.0f];
    self.userAvatarImageView.hidden = YES;
    self.videoContainerView.hidden = NO;
    self.unmuteOnVideoImageView.hidden = YES;
    self.nameLabelCenterXConstraint.constant = 0.0f;
    UIPinchGestureRecognizer *pinchGesture = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(pinchGesture:)];
    [self.videoContainerView addGestureRecognizer:pinchGesture];
}

- (void)setIsResizeAspect:(BOOL)isResizeAspect {
    if (_isResizeAspect != isResizeAspect) {
        _isResizeAspect = isResizeAspect;
        if (self.didChangeVideoGravity) {
             self.didChangeVideoGravity(isResizeAspect);
        }
    }
}

- (void)pinchGesture:(UIPinchGestureRecognizer *)gestureRecognizer {
    if (gestureRecognizer.view != nil) {
        if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {
            CGFloat currentScale = self.videoContainerView.frame.size.width / self.videoContainerView.bounds.size.width;
            CGFloat newScale = currentScale * gestureRecognizer.scale;
            if (self.isResizeAspect == YES) {
                if (newScale < currentScale) {
                    self.isResizeAspect = NO;
                }
            } else {
                if (newScale > currentScale) {
                    self.isResizeAspect = YES;
                }
            }
        }
    }
}

- (void)setVideoView:(UIView *)videoView {
    
    _videoView = videoView;
    
    [self.videoContainerView insertSubview:videoView atIndex:0];
        
    videoView.translatesAutoresizingMaskIntoConstraints = NO;
    [videoView.leftAnchor constraintEqualToAnchor:self.videoContainerView.leftAnchor].active = YES;
    [videoView.rightAnchor constraintEqualToAnchor:self.videoContainerView.rightAnchor].active = YES;
    [videoView.topAnchor constraintEqualToAnchor:self.videoContainerView.topAnchor].active = YES;
    [videoView.bottomAnchor constraintEqualToAnchor:self.videoContainerView.bottomAnchor].active = YES;
    
    [videoView layoutIfNeeded];
}

- (void)setName:(NSString *)name {
    if (!name) {
        return;
    }
    if (![_name isEqualToString:name]) {
        _name = [name copy];
        
        self.userNameTopLabel.text = self.name;
        self.nameLabel.text = self.name;
        self.userAvatarLabel.text = name.firstLetter;
    }
}

- (void)setNameColor:(UIColor *)nameColor {
    if (![_nameColor isEqual:nameColor]) {
        _nameColor = nameColor;
        self.userAvatarLabel.backgroundColor = nameColor;
    }
}

- (void)setIsMuted:(BOOL)isMuted {
    if (_isMuted != isMuted) {
        _isMuted = isMuted;
        self.unmuteImageViewWidthConstraint.constant = isMuted ? 0.0f : 40.0f;
        self.nameLabelCenterXConstraint.constant = isMuted ? 0.0f : -10.0f;
        self.unmuteOnVideoImageView.hidden = isMuted;
    }
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/ConferenceUserCel/ConferenceUserCell.swift
//
//  ConferenceUserCell1.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 17.08.2020.
//  Copyright © 2020 QuickBlox. All rights reserved.
//

import UIKit

struct ConferenceUserCellConstant {
   static let reuseIdentifier = "ConferenceUserCell"
}

class ConferenceUserCell: UICollectionViewCell {

       //MARK: - IBOutlets
        @IBOutlet private weak var userView: UIView!
        @IBOutlet private weak var userAvatarLabel: UILabel!
        @IBOutlet private weak var userAvatarImageView: UIImageView!
        @IBOutlet weak var userNameTopLabel: UILabel!
        @IBOutlet private weak var nameLabel: UILabel!
        @IBOutlet private weak var containerView: UIView!
        @IBOutlet weak var unmuteImageViewWidthConstraint: NSLayoutConstraint!
        @IBOutlet weak var unmuteImageView: UIImageView!
        @IBOutlet weak var nameLabelCenterXConstraint: NSLayoutConstraint!
        @IBOutlet weak var videoContainerView: UIView!
        @IBOutlet weak var unmuteOnVideoImageView: UIImageView!
        
        //MARK: - Properties
        /**
         *  Change Video Gravity block action.
         */
        var didChangeVideoGravity: ((_ isResizeAspect: Bool) -> Void)?
        
        var videoView: UIView? {
            didSet {
                guard let view = videoView else {
                    return
                }
                videoContainerView.insertSubview(view, at: 0)
                view.translatesAutoresizingMaskIntoConstraints = false
                view.leftAnchor.constraint(equalTo: videoContainerView.leftAnchor).isActive = true
                view.rightAnchor.constraint(equalTo: videoContainerView.rightAnchor).isActive = true
                view.topAnchor.constraint(equalTo: videoContainerView.topAnchor).isActive = true
                view.bottomAnchor.constraint(equalTo: videoContainerView.bottomAnchor).isActive = true
                view.layoutIfNeeded()
            }
        }
        
        var unMute = false {
            didSet {
                unmuteImageViewWidthConstraint.constant = unMute == true ? 0.0 : 40.0
                nameLabelCenterXConstraint.constant = unMute == true ? 0.0 : -10.0
                unmuteOnVideoImageView.isHidden = unMute == true
            }
        }
        
        var videoEnabled = false {
            didSet {
                videoContainerView.isHidden = !videoEnabled
            }
        }
        
        var name = "" {
            didSet {
                userNameTopLabel.text = name
                nameLabel.text = name
                userAvatarLabel.text = String(name.capitalized.first ?? Character("Q"))
            }
        }
        
        var userColor = UIColor.clear {
            didSet {
                userAvatarLabel.backgroundColor = userColor
            }
        }
        
        var isResizeAspect = true {
            didSet {
                didChangeVideoGravity?(isResizeAspect)
            }
        }
        
        //MARK: - Life Cycle
        override func awakeFromNib() {
            super.awakeFromNib()
            
            backgroundColor = .clear
            videoContainerView.backgroundColor = .clear
            containerView.backgroundColor = #colorLiteral(red: 0.1960526407, green: 0.1960932612, blue: 0.1960500479, alpha: 1)
            userAvatarLabel.setRoundedLabel(cornerRadius: 30.0)
            userAvatarImageView.setRoundedView(cornerRadius: 30.0)
            userAvatarImageView.isHidden = true
            videoContainerView.isHidden = true
            unmuteOnVideoImageView.isHidden = true
            nameLabelCenterXConstraint.constant = 0.0
            let pinchGesture = UIPinchGestureRecognizer(target: self, action: #selector(pinchGesture(_ :)))
            videoContainerView.addGestureRecognizer(pinchGesture)
        }
    
    //MARK: - Actions
    @objc private func pinchGesture(_ gestureRecognizer: UIPinchGestureRecognizer) {
        guard gestureRecognizer.view != nil else { return }
        if gestureRecognizer.state != .began {
            return
        }
        let currentScale = videoContainerView.frame.width / videoContainerView.bounds.width
        let newScale = currentScale * gestureRecognizer.scale
        if (isResizeAspect == true && newScale < currentScale)
            || (isResizeAspect == false && newScale > currentScale) {
            isResizeAspect = !isResizeAspect
        }
    }
}
